src/relay/backend/interpreter.cc

```diff
@@ -283,73 +290,153 @@ class Interpreter : public ExprFunctor<ObjectRef(const Expr& n)>,
+    // itself. We currently can't express shape function types in Relay.
+    const FuncTypeNode* ftn = prim_shape_fn_var->checked_type().as<FuncTypeNode>();
+    ICHECK(ftn);
+    ICHECK_EQ(prim_shape_fn_states.size(), ftn->arg_types.size());
+    ICHECK_EQ(args.size(), ftn->arg_types.size());
+    // num_shape_inputs will account for which primitive function arguments are dynamic,
+    // whether the shape and or data needs to be passed, and flattening of ADTs.
+    // Similarly, num_shape_outputs will account for flattening of ADTs.
+
+    PackedFunc packed_shape_func = Build(prim_shape_fn_var, all_prim_shape_fn_vars);
+
+    size_t arity = num_shape_inputs + num_shape_outputs;
     std::vector<TVMValue> values(arity);
     std::vector<int> codes(arity);
     TVMArgsSetter setter(values.data(), codes.data());
-    std::vector<NDArray> inputs(cfunc->inputs.size());
-    std::vector<NDArray> outputs(cfunc->outputs.size());
+    std::vector<NDArray> inputs(num_shape_inputs);
+    std::vector<NDArray> outputs(num_shape_outputs);
 
     Device cpu_dev;
     cpu_dev.device_type = kDLCPU;
     cpu_dev.device_id = 0;
 
-    auto fset_input = [&](size_t i, ObjectRef val, bool need_shape) {
+    auto fset_shape_input = [&](size_t i, ObjectRef val) {
+      DCHECK(i < num_shape_inputs);
       auto nd_array = Downcast<NDArray>(val);
-      if (need_shape) {
-        int64_t ndim = nd_array.Shape().size();
-        NDArray shape_arr;
-        if (ndim == 0) {
-          shape_arr = NDArray::Empty({}, DataType::Int(64), cpu_dev);
-        } else {
-          shape_arr = NDArray::Empty({ndim}, DataType::Int(64), cpu_dev);
-          int64_t* data = reinterpret_cast<int64_t*>(shape_arr->data);
-          for (auto j = 0; j < ndim; ++j) {
-            data[j] = nd_array.Shape()[j];
-          }
-        }
-        inputs[i] = shape_arr;
-        setter(i, shape_arr);
+      int64_t ndim = nd_array.Shape().size();
+      NDArray shape_arr;
+      if (ndim == 0) {
+        shape_arr = NDArray::Empty({}, DataType::Int(64), cpu_dev);
       } else {
-        auto arr = nd_array.CopyTo(cpu_dev);
-        inputs[i] = arr;
-        setter(i, arr);
+        shape_arr = NDArray::Empty({ndim}, DataType::Int(64), cpu_dev);
+        int64_t* data = reinterpret_cast<int64_t*>(shape_arr->data);
+        for (auto j = 0; j < ndim; ++j) {
+          data[j] = nd_array.Shape()[j];
+        }
       }
+      inputs[i] = shape_arr;
+      setter(i, shape_arr);
+    };
+
+    auto fset_data_input = [&](size_t i, ObjectRef val) {
+      DCHECK(i < num_shape_inputs);
+      auto nd_array = Downcast<NDArray>(val);
+      auto arr = nd_array.CopyTo(cpu_dev);
+      inputs[i] = arr;
+      setter(i, arr);
     };
 
     size_t arg_counter = 0;
     for (size_t i = 0; i < args.size(); ++i) {
       auto arg = args[i];
-      auto param = func->params[i];
-      int state = cfunc->shape_func_param_states[i]->value;
+      int64_t state = prim_shape_fn_states[i]->value;
```

Comment:
This kNeedInputData/kNeedInputShape state per arg is a bit gnarly since it's gathered as a side-effort of building the shape function. So I'm gonna leave a TODO to cleanup the way we convey this to call sites rather than kick that hornets nest right now.

Done for TupleType flattening.

**However** is ADT::Tuple coherent with FlattenTupleType? Ie is this legit?
```
std::vector<TensorType> result_tensor_types = FlattenTupleType(ftn->ret_type);
std::vector<ObjectRef> result_ndarrays;
for (int i = 0; i < result_tensor_types.size(); ++i) {
   ...
   result_ndarrays.emplace_back(...)
}
if (result_tensor_types.size() == 1) {
      return result_ndarrays[0];
} else {
  return ADT::Tuple(result_ndarrays);
}
```

Reply:
I ended up redoing the ADT vs NDArray stuff to mirror the FlattenTupleType and friends. Much cleaner. But for nested tuples there's going to be mismatches until we get this cleaned up throughout the codebase.